src/libraries/Common/src/System/Collections/Generic/ArrayBuilder.cs

```diff
@@ -70,6 +70,39 @@ public void Add(T item)
             UncheckedAdd(item);
         }
 
+        /// <summary>
+        /// Adds a range of items to the backing array, resizing it as necessary.
+        /// </summary>
+        /// <param name="items">The items to add.</param>
+        public void AddRange(IEnumerable<T> items)
+        {
+            Debug.Assert(items != null);
+
+            if (items is ICollection<T> collection)
+            {
+                int count = collection.Count;
+                if (count > 0)
+                {
+                    if (_count + count > Capacity)
+                    {
+                        EnsureCapacity(_count + count);
+                    }
+
+                    foreach (T item in collection)
+                    {
+                        UncheckedAdd(item);
+                    }
```

Comment:
This is a deoptimization over using List, as it's copying element by element via enumeration rather than using ICollection.CopyTo.